crates/anvil-polkadot/src/substrate_node/mining_engine.rs

```diff
@@ -0,0 +1,344 @@
+    interval_ticker.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+    let stream = unfold(interval_ticker, |mut interval_tick| async {
+        interval_tick.tick().await;
+        Some(((), interval_tick))
+    });
+    Box::pin(stream.fuse())
+}
+
+fn build_auto_stream(engine: &Arc<MiningEngine>) -> SealCommandStream {
+    let stream = engine.transaction_pool.import_notification_stream().map(|_| ());
+    Box::pin(stream.fuse())
+}
+
+fn build_streams_for_mode(
+    mode: MiningMode,
+    engine: &Arc<MiningEngine>,
+) -> (Option<SealCommandStream>, Option<SealCommandStream>) {
+    let interval_stream = match mode {
+        MiningMode::Interval { tick } | MiningMode::MixedMining { tick } => Some(tick),
+        _ => None,
+    }
+    .map(build_interval_stream);
+    let auto_stream = matches!(mode, MiningMode::AutoMining | MiningMode::MixedMining { .. })
+        .then(|| build_auto_stream(engine));
+    (interval_stream, auto_stream)
+}
+
+async fn wait_for_mode_change(
+    engine: &Arc<MiningEngine>,
+    current: Option<MiningMode>,
+) -> MiningMode {
+    futures::future::poll_fn(|cx| {
+        let mode = *engine.mining_mode.read();
+        if current.as_ref().is_none_or(|m| *m != mode) {
+            return std::task::Poll::Ready(mode);
+        }
+        engine.waker.register(cx.waker());
+        std::task::Poll::Pending
+    })
+    .await
+}
+
+async fn next_or_pending(stream: Option<&mut SealCommandStream>) -> Option<()> {
+    match stream {
+        Some(s) => s.next().await,
+        None => futures::future::pending().await,
+    }
+}
+
+async fn poll_and_seal(engine: Arc<MiningEngine>, stream: Option<&mut SealCommandStream>) -> bool {
+    if next_or_pending(stream).await.is_some() {
+        match engine.seal_now().await {
+            Ok(block) => {
+                debug!(hash=?block.hash, "sealed");
+            }
+            Err(BlockProducingError::Canceled(_) | BlockProducingError::SendError(_)) => {
+                return false; // fatal: break outer loop
+            }
+            Err(e) => {
+                error!(?e, "block production failed");
+            }
+        }
+    }
+    true
+}
+
+pub async fn run_mining_engine(engine: Arc<MiningEngine>) {
+    let mut current_mode = None;
+    let mut interval_mining_stream: Option<SealCommandStream> = None;
+    let mut auto_mining_stream: Option<SealCommandStream> = None;
+
+    loop {
+        tokio::select! {
+            new_mode = wait_for_mode_change(&engine, current_mode) => {
+                current_mode = Some(new_mode);
+                let (interval_stream, auto_stream) = build_streams_for_mode(new_mode, &engine);
+                interval_mining_stream = interval_stream;
+                auto_mining_stream = auto_stream;
+            }
+            // Poll the interval stream if it exists
+            continue_loop = poll_and_seal(engine.clone(), interval_mining_stream.as_mut()) => {
```

Comment:
By guarding this branch you can get rid of `next_or_pending`.
```suggestion
            continue_loop = poll_and_seal(engine.clone(), interval_mining_stream.as_mut()), if  interval_mining_stream.is_some() => {
```